Decrease internal descriptor counter of the DescriptorAllocator on free

hasenbanck opened the pull request:
While debugging an out of memory error in the WGPU Vulkan backend ([issue](https://github.com/gfx-rs/wgpu/issues/7525)) I realized, that `gpu-descriptor` is never actually decreasing the internal `DescriptorAllocator::total` and `DescriptorAllocator::current_update_after_bind_descriptors_in_all_pools` counts. It only seems to increase the counters on allocation, but the deallocation does not decrease the counter of the `DescriptorAllocator`.

With large descritptor sets (when doing bindless), this will result in OOM errors in a couple of seconds.

I verified my changes against the current Vulkan validation layers.

zakarumych commented:
Ah, I see. This was clearly an oversight.
I've never encountered this issue due to allocating up-front instead of continuous allocation-deallocation cycles.

hasenbanck commented:
Are there any issues left getting this PR merged?

zakarumych commented:
I think this should work. But to be on the safe side I'd move subtraction to the line after actual set deallocation.
It happens when `bucket.free` is called. Up to this point, you may collect descriptors count into local variable. Once `bucket.free` returns, move this count to the `self.total` and `self.current_update_after_bind_descriptors_in_all_pools`.

hasenbanck committed:
Decrease internal descriptor counter of the DescriptorAllocator on free

hasenbanck committed:
Fix linter issues

Because we use "is_some_and()" now, this increases the MSRV to 1.70. The previous MSRV seems to have been 1.69 when using ash.

hasenbanck commented:
I tried to implement your proposed changes and also then decided to maybe include a helper function to reduce code duplication. Since the GithHub Action check for the linter threw errors, I took the oppotunity to fix them in a seperate commit, which we could also drop if you want, since those fixes would increase the MSRV to 1.70. The current MSRV when using ash is 1.69.

zakarumych commented:
Thank you!

1.70 is fine. I'll go ahead and merge this and prepare a patch version

hasenbanck commented:
> Thank you!
> 
> 1.70 is fine. I'll go ahead and merge this and prepare a patch version

Could you please release a patch version to make this fix available upstream (wgpu)?

zakarumych commented:
It is done
